### PR TITLE
Add CRD distillation strategy and tests

### PR DESCRIPTION
**Pull request by surajyadav-research**
## Summary
Adds **Contrastive Representation Distillation (CRD)** to Tunix: an InfoNCE-style contrastive loss between student/teacher **intermediate representations** captured via `sowed_module`, combined with the supervised loss. #887

## Implementation (CRD)
- Wraps selected student/teacher layers with `sowed_module` to capture intermediates.
- Pools each captured feature leaf to `(B, C)`:
  - `(B, C)` → use as-is  
  - `(B, T, C)` + `input_mask` → masked mean over `T`  
  - otherwise mean over non-batch/non-channel dims (e.g. `(B, H, W, C)`).
- Stacks pooled leaves to `(N, B, C)` and averages over `N` to get `(B, C)`.
- Projects to a shared embedding space (MLP/linear), L2-normalizes, and computes **InfoNCE with in-batch negatives** (optional symmetric).
- Training loss: `alpha * crd_loss + (1 - alpha) * task_ce`. Teacher features use `stop_gradient`.
- `pre_process_models` wraps models + adds heads; `post_process_models` unwraps and restores original models.

## Tests
- `feature_extraction_test.py`
  - Avg-pooling (VALID/SAME + include/exclude pad) + error cases
  - `sowed_module` wrap/pop/unwrap (single + multi-leaf capture; shape-safe matching)
  - Feature projection setup/removal integration
- `strategies_test.py`
  - LogitStrategy matches manual KL+CE
  - Smoke tests: FeaturePoolingStrategy, FeatureProjectionStrategy, and CRD
  - Verifies losses are scalars and sowed wrappers are removed in post-process

## Testing
```bash
python -m ruff format --check \
  tunix/distillation/strategies/crd_strategy.py \
  tests/distillation/feature_extraction_test.py \
  tests/distillation/strategies_test.py

python -m pytest -q \
  tests/distillation/feature_extraction_test.py \
  tests/distillation/strategies_test.py

### CODE REVIEW

**Comment by surajyadav-research** [comment]
Hi @abheesht17 @hgao327,
When you get a chance, could you please review this code? I’ll update it according to your feedback.
